Source/AcceleratorLattice/LatticeElements/Solenoid.H

```diff
@@ -0,0 +1,164 @@
+    std::vector<std::vector<amrex::ParticleReal>> h_b_coef;
+
+    // Device variable read from input files
+    amrex::Gpu::DeviceVector<amrex::ParticleReal> d_scale;
+    amrex::Gpu::DeviceVector<amrex::ParticleReal> d_freq;
+    amrex::Gpu::DeviceVector<amrex::ParticleReal> d_theta;
+    amrex::Vector<amrex::Gpu::DeviceVector<amrex::ParticleReal>> d_e_coef;
+    amrex::Vector<amrex::Gpu::DeviceVector<amrex::ParticleReal>> d_b_coef;
+
+    /**
+     * \brief Returns the device level instance with the lattice information
+     */
+    SolenoidDevice GetDeviceInstance () const;
+
+
+};
+
+// Instance that is trivially copyable to the device.
+
+struct SolenoidDevice
+{
+
+    /**
+     * \brief Initializes the data and pointer needed to reference the lattice element info
+     *
+     * @param[in] h_solenoid host level instance that this is associated with
+     */
+    void InitSolenoidDevice (Solenoid const& h_solenoid);
+
+    int nelements = 0;
+
+    const amrex::ParticleReal* AMREX_RESTRICT d_zs_arr;
+    const amrex::ParticleReal* AMREX_RESTRICT d_ze_arr;
+    const amrex::ParticleReal* AMREX_RESTRICT d_scale_arr;
+    const amrex::ParticleReal* AMREX_RESTRICT d_freq_arr;
+    const amrex::ParticleReal* AMREX_RESTRICT d_theta_arr;
+    unsigned long* AMREX_RESTRICT d_num_e_coef;
+    unsigned long* AMREX_RESTRICT d_num_b_coef;
+    const amrex::ParticleReal** AMREX_RESTRICT d_e_coef_arr;
+    const amrex::ParticleReal** AMREX_RESTRICT d_b_coef_arr;
+
+    /**
+     * \brief Fetch the field of the specified element at the given location
+     *
+     * @param[in] ielement the element number
+     * @param[in] x, y, z the particle position in the lab frame
+     * @param[in] m_time the current time in the lab frame
+     * @param[out] Ex, Ey, Ez, Bx, By, Bz the fetched field in the lab frame
+     */
+    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+    void get_field (const int ielement,
+                    const amrex::ParticleReal x,
+                    const amrex::ParticleReal y,
+                    const amrex::ParticleReal z,
+                    const amrex::ParticleReal m_time,
+                    amrex::ParticleReal& Ex,
+                    amrex::ParticleReal& Ey,
+                    amrex::ParticleReal& Ez,
+                    amrex::ParticleReal& Bx,
+                    amrex::ParticleReal& By,
+                    amrex::ParticleReal& Bz) const
+    {
+
+        const amrex::ParticleReal zs = d_zs_arr[ielement];
+        const amrex::ParticleReal ze = d_ze_arr[ielement];
+        const amrex::ParticleReal scale = d_scale_arr[ielement];
+        const amrex::ParticleReal freq = d_freq_arr[ielement];
+        const amrex::ParticleReal theta = d_theta_arr[ielement];
+        const amrex::ParticleReal* e_coef = d_e_coef_arr[ielement];
+        const amrex::ParticleReal* b_coef = d_b_coef_arr[ielement];
+
+        const amrex::Real pi = 2.0*asin(1.0);
+        const amrex::Real c = 299792458.0;
```

Comment:
For consistency with other code, these should be
```
amrex::ParticleReal constexpr pi = MathConst::pi;
amrex::ParticleReal constexpr c = PhysConst::c;
```

Reply:
@dbizzoze the header for this is:
```C++
#include "Utils/WarpXConst.H"
```